### PR TITLE
Enhance analyze command API to understand modules' needs

### PR DESCRIPTION
**Pull request by acammack-r7**
Beginning steps to enhance the `analyze` command to be able to understand what potentially testable modules will need in order to run. Currently re-working vuln/service matching to lose less information. Putting more module option information in the module cache and/or having a step that instantiates good matches will likely be needed soon to provide better targeting.

### CODE REVIEW

**Comment by pbarry-r7** [comment]
Added folks as reviewers for visibility that this is posted up.  Thanks!

File: lib/msf/ui/console/command_dispatcher/db/analyze.rb [hunk 1]
```diff
@@ -44,23 +47,23 @@ def cmd_analyze(*args)
       host_ids.each do |id|
```
Comment (adfoster-r7) [inline lib/msf/ui/console/command_dispatcher/db/analyze.rb:47]: Not a blocker for this PR, but is it worth improving the logging for the scenario where there's hosts available but nothing is suggested?

```
msf6 auxiliary(scanner/smb/smb_version) > hosts

Hosts
=====

address          mac                name            os_name       os_flavor  os_sp  purpose  info  comments
-------          ---                ----            -------       ---------  -----  -------  ----  --------
192.168.18.118   00:00:00:00:00:00  VAGRANT-2008R2
192.168.222.151  00:00:00:00:00:00                  Windows 2008                    server

msf6 auxiliary(scanner/smb/smb_version) > analyze
msf6 auxiliary(scanner/smb/smb_version) > analyze
msf6 auxiliary(scanner/smb/smb_version) > 
```

I initially thought the command was breaking silently when it gave no output 😄

File: spec/lib/msf/core/analyze_spec.rb [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+    let(:ref_3) { FactoryBot.create(:mdm_ref) }
+    let(:ref_4) { FactoryBot.create(:mdm_ref) }
+
+    let(:vuln_a) { FactoryBot.create(:mdm_vuln) }
+    let(:vuln_ap1) { FactoryBot.create(:mdm_vuln) }
+    let(:vuln_b) { FactoryBot.create(:mdm_vuln) }
+    let(:vuln_c) { FactoryBot.create(:mdm_vuln) }
+    let(:vuln_cta) { FactoryBot.create(:mdm_vuln) }
+    let(:vuln_d) { FactoryBot.create(:mdm_vuln) }
+    let(:vuln_dtc) { FactoryBot.create(:mdm_vuln) }
```
[the remark this answers is not in the document]
Comment (acammack-r7) [reply spec/lib/msf/core/analyze_spec.rb:17]: I'm open to it, but subject names like `c_transitive_to_a` didn't seem to read as well in the tests. If you think that would work better I can update the tests.

File: lib/msf/core/db_manager/import.rb [hunk 3]
```diff
@@ -136,7 +135,7 @@ def import(args={}, &block)
             next
           end
           next unless mports[service.port]
-          if (match_names - mports[service.port].keys).count < match_names.count
+          if (match_vulns - mports[service.port]).size < matched_vulns.size
```
[the remark this answers is not in the document]
Comment (acammack-r7) [reply lib/msf/core/db_manager/import.rb:138]: Good catch, thanks

File: lib/msf/ui/console/command_dispatcher/db/analyze.rb (hunk 1 above)
Comment (acammack-r7) [reply lib/msf/ui/console/command_dispatcher/db/analyze.rb:47]: Yeah, that's existing behaviour, but it seems worthwhile to fix. Probably worth having separate messages for no hosts, hosts with no vulns, and hosts with vulns that don't match anything.

File: lib/msf/core/analyze.rb [hunk 4]
```diff
@@ -5,85 +5,106 @@ def initialize(framework)
...
+            # Whoever was initialized first has the canonical set
+            grouped_vulns[candidate_match] = grouped_vulns[vuln]
+          end
+        end
+      end
+    end
 
-    false
+    vuln_families = grouped_vulns.values
+    vuln_families.uniq! || vuln_families
```
Comment (adfoster-r7) [inline lib/msf/core/analyze.rb:59]:
```suggestion
    vuln_families.uniq!
    vuln_families
```

File: lib/msf/core/analyze/result.rb [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+      return false if set.include?("aix") && mod.refname.include?("aix") && !hos.include?("aix")
+      # Skip IRIX bugs if we have a solid match against another OS
+      return false if set.include?("irix") && mod.refname.include?("irix") && !hos.include?("irix")
+
+      return true if !hos.include?('windows')
+    end
+
+    return true if set.include?("php")
+    return true if set.include?("python")
+    return true if set.include?("java")
```
[the remark this answers is not in the document]
Comment (acammack-r7) [reply lib/msf/core/analyze/result.rb:154]: No, those platforms are only used by generic encoders and payloads.

File: spec/lib/msf/core/analyze_spec.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (adfoster-r7) [reply spec/lib/msf/core/analyze_spec.rb:17]: Unpacking the acronym seems like a good shout :+1: My brain automatically converted `cta` into `call to action` 😄

**Comment by adfoster-r7** [comment]
Looks good to me :+1:

I can get this landed after the those last tweaks are in 🚢  :shipit:

File: lib/msf/core/analyze/result.rb [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+    end
+
+    if @mod.post_auth?
+      unless @mod.default_cred? || have_service_cred? || have_datastore_cred?
+        @missing << :credential
+      end
+    end
+  end
+
+  def determine_prerequisites
```
[the remark this answers is not in the document]
Comment (acammack-r7) [reply lib/msf/core/analyze/result.rb:69]: Currently, the only intended use for `prerequisites_evaluated` is determining `ready_for_test?`. If we need to short-circuit evaluating with the same datastore multiple times we can add some sort of guard here later.

File: lib/msf/core/db_manager/import.rb [hunk 7]
```diff
@@ -146,7 +145,7 @@ def import(args={}, &block)
         if serv.nil? && !second_pass_services.empty?
           second_pass_services.each do |service|
             next unless mports[service.port]
-            if (match_names - mports[service.port].keys).count < match_names.count
+            if (match_vulns - mports[service.port]).size < matched_vulns.size
```
Comment (adfoster-r7) [inline lib/msf/core/db_manager/import.rb:148]: I think just this line just needs the same rename applied like the above line, then I think that's everything to get this landed 🎉

**Comment by adfoster-r7** [comment]
Imports and analyzing working:
```
msf6 > db_import ~/Downloads/m3_report.xml
[*] Successfully imported /Users/adfoster/Downloads/m3_report.xml
msf6 > analyze
[*] Analysis for 192.168.18.118 ->
[*]   exploit/windows/smb/ms17_010_eternalblue - ready for testing
[*]   exploit/windows/smb/ms17_010_eternalblue_win8 - ready for testing
[*]   exploit/windows/smb/ms17_010_psexec - credentials are required
[*]   exploit/windows/smb/smb_doublepulsar_rce - ready for testing
msf6 > exit
```

And sweet bonus UX improvements too 🎉 🙇‍♂️ 
```
msf6 > analyze
[*] No existing hosts stored to analyze.
```

With hosts but no vulns:
```
msf6 > analyze
[*] No matching modules found.
```

Also confirmed what was discussed during module hacking - the speed is acceptable for a small number of hosts/vulns, but might need some tweaks to scale to larger data sets - but overall a great first step :+1:
```
msf6 exploit(multi/http/cockpit_cms_rce) > time analyze 192.168.18.118
[*] Analysis for 192.168.18.118 ->
[*]   exploit/windows/smb/ms17_010_eternalblue - ready for testing
[*]   exploit/windows/smb/ms17_010_eternalblue_win8 - ready for testing
[*]   exploit/windows/smb/ms17_010_psexec - credentials are required
[*]   exploit/windows/smb/smb_doublepulsar_rce - ready for testing
[+] Command "analyze 192.168.18.118" completed in 1.021813000086695 seconds
```

I think it's just that last line that needs a tweak then we can get this shipped 🚢

**Comment by adfoster-r7** [comment]
## Release Notes

Enhances the analyze command to show additional information about an identified exploit being immediately runnable, or if it requires additional credentials or options to be set before being ran

**Pull request merged**